### PR TITLE
Indexing of parameters

### LINKED ISSUE #231: Integrator symbolic indexing should support parameters
`getindex(A::DEIntegrator, sym)` can flexibly return states, independent variable, and observed variables, but not system parameters. 

I would like subtypes of `AbstractDiffEqFunction{iip}` (those used by MTK) to have a kwarg and field equivalent to `syms`, called `paramsyms`. That would allow modifying the `getindex` method to support all possible system symbols. The changes here would need to be propagated to the contructor calls in `ModelingToolkit.jl`

https://github.com/SciML/SciMLBase.jl/blob/f348d7c377aaa8006069971f5dc3bba6b89fabce/src/integrator_interface.jl#L415-L437

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Closes #231

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
Is there any particular reason several `*DEFunction` don't have `indepsym`? For example, `SplitFunction`, `DynamicalODEFunction`, `DDEFunction`. They all mention it in the docstrings, which I guess is an artifact of copying the same docstring in multiple places.

**Comment by AayushSabharwal** [comment]
Put it back as draft since there is some weirdness going on

**Comment by AayushSabharwal** [comment]
I wish merge conflicts weren't a thing

**Comment by AayushSabharwal** [comment]
_I just fixed these for the second time_

**Comment by AayushSabharwal** [comment]
Thank god

**Comment by AayushSabharwal** [comment]
It's a bit of a chicken-and-egg situation here. The test won't pass because MTK doesn't actually use the new keyword argument to give it the parameter symbols, but to make the changes in MTK this needs to be merged.

**Comment by ChrisRackauckas** [comment]
Finish the MTK part first.

File: src/scimlfunctions.jl [hunk 1]
```diff
@@ -266,7 +271,7 @@ SplitFunction{iip,recompile}(f1,f2;
                              sparsity = __has_sparsity(f) ? f.sparsity : jac_prototype,
                              paramjac = __has_paramjac(f) ? f.paramjac : nothing,
                              syms = __has_syms(f) ? f.syms : nothing,
-                             indepsym= __has_indepsym(f) ? f.indepsym : nothing,
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:269]: why was this removed? This seems like a mistake?

File: src/scimlfunctions.jl [hunk 2]
```diff
@@ -895,7 +914,7 @@ SplitSDEFunction{iip,recompile}(f1,f2,g;
                  sparsity = __has_sparsity(f) ? f.sparsity : jac_prototype,
                  paramjac = __has_paramjac(f) ? f.paramjac : nothing,
                  syms = __has_syms(f) ? f.syms : nothing,
-                 indepsym= __has_indepsym(f) ? f.indepsym : nothing,
+                 paramsyms
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:917]: typo?

File: src/scimlfunctions.jl [hunk 3]
```diff
@@ -1010,7 +1031,7 @@ DynamicalSDEFunction{iip,recompile}(f1,f2;
                                     sparsity = __has_sparsity(f) ? f.sparsity : jac_prototype,
                                     paramjac = __has_paramjac(f) ? f.paramjac : nothing,
                                     syms = __has_syms(f) ? f.syms : nothing,
-                                    indepsym= __has_indepsym(f) ? f.indepsym : nothing,
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:1013]: typo?

**Comment by ChrisRackauckas** [comment]
Looks like there's a few typos to clean up, and some conflicts (please rebase), and then this is good to go

File: src/integrator_interface.jl [hunk 4]
```diff
@@ -423,9 +431,12 @@ Base.@propagate_inbounds function Base.getindex(A::DEIntegrator, sym)
     end
 
     indepsym = getindepsym(A)
+    paramsyms = getparamsyms(A)
     if i === nothing
         if issymbollike(sym) && indepsym !== nothing && Symbol(sym) == indepsym
             A.t
+        elseif issymbollike(sym) && paramsyms !== nothing && Symbol(sym) in paramsyms
+            paramsyms[findfirst(Symbol(sym), paramsyms)]
```
[the remark this answers is not in the document]
Comment (wsphillips) [reply src/integrator_interface.jl:439]: (and similarly on solution interface)

[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/integrator_interface.jl:439]: Seems like the test should be improved to catch this. Instead of just indexing, test the value that comes out too.

**Comment by AayushSabharwal** [comment]
> Is there any particular reason several `*DEFunction` don't have `indepsym`? For example, `SplitFunction`, `DynamicalODEFunction`, `DDEFunction`. They all mention it in the docstrings, which I guess is an artifact of copying the same docstring in multiple places.

I removed `indepsym` from the docstrings because it wasn't used in the corresponding constructors or present in the structs

**Comment by ChrisRackauckas** [comment]
That sounds backwards though: any time-dependent equation should have an indepsym?

**Comment by AayushSabharwal** [comment]
It should, I'm not sure why they don't. Should I add them? Both here and in MTK

**Comment by ChrisRackauckas** [comment]
Yes

File: test/downstream/symbol_indexing.jl [hunk 5]
```diff
@@ -66,100 +66,102 @@ sol = solve(prob, Rodas4())
...
-eqs = [collect(D.(x) .~ x)
-       D(y) ~ norm(x) * y - x[1]]
-@named sys = ODESystem(eqs, t, [sts...;], [ps...;])
-prob = ODEProblem(sys, [], (0, 1.0))
-@test_broken sol = solve(prob, Tsit5())
-@test_broken sol[x] isa Vector{<:Vector}
-@test_broken sol[@nonamespace sys.x] isa Vector{<:Vector}
+# interpolated_sol = sol(0.0:1.0:10.0)
+# @test interpolated_sol[α] isa Vector
+# @test interpolated_sol[α, :] isa Vector
```
Comment (ChrisRackauckas) [inline test/downstream/symbol_indexing.jl:75]: Never ever comment out tests like this.

Why are these commented?

Comment (AayushSabharwal) [reply test/downstream/symbol_indexing.jl:75]: Ah, my bad I didn't mean to commit this. I do this to selectively run tests locally

**Comment by AayushSabharwal** [comment]
Working on rebasing, this is gonna take a while

**Comment by AayushSabharwal** [comment]
Turns out I'm _really_ bad at rebasing

File: src/SciMLBase.jl [hunk 6]
```diff
@@ -568,8 +568,6 @@ function specialization(f::Union{ODEFunction{iip, specialize},
     specialize
 end
 
-specialization(f::AbstractSciMLFunction) = FullSpecialize
```
Comment (ChrisRackauckas) [inline src/SciMLBase.jl:571]: why?

File: src/problems/ode_problems.jl [hunk 7]
```diff
@@ -105,31 +105,22 @@ struct ODEProblem{uType, tType, isinplace, P, F, K, PT} <:
...
         ptspan = promote_tspan(tspan)
-        _f = ODEFunction{iip, AutoSpecialize}(f)
+        _f = if iip && typeof(u0) <: Vector{Float64} &&
+                eltype(promote_tspan(tspan)) <: Float64 &&
+                typeof(p) <: Union{Vector{Float64}, NullParameters}
+            ff = wrapfun_iip(f, (u0, u0, p, ptspan[1]))
+            ODEFunction{iip, AutoSpecialize}(ff)
+        else
+            ODEFunction{iip}(f)
+        end
```
Comment (ChrisRackauckas) [inline src/problems/ode_problems.jl:115]: Unrelated?

**Comment by AayushSabharwal** [comment]
That might be what happened. I don't know why it kept finding conflicts in files I haven't touched, and apparently I resolved them wrong. I'll try fixing this, and if I can't do so in a reasonable amount of time I think I'll just create a new branch and make the changes there

File: src/scimlfunctions.jl [hunk 8]
```diff
@@ -2948,18 +3465,46 @@ function OptimizationFunction{iip}(f, adtype::AbstractADType = NoAD();
...
+<<<<<<< HEAD
                                       cons_hess_prototype, lag_hess_prototype, syms,
                                       hess_colorvec,
                                       cons_jac_colorvec, cons_hess_colorvec,
                                       lag_hess_colorvec, expr,
                                       cons_expr, sys)
+=======
+                                      cons_hess_prototype, syms, paramsyms,
+=======
+                                      cons_hess_prototype, syms, indepsym, paramsyms,
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:3500]: This is a merge error.

File: src/scimlfunctions.jl [hunk 9]
```diff
@@ -2802,18 +3292,43 @@ function SDDEFunction{iip, recompile}(f, g;
...
+                     typeof(paramsyms), typeof(observed),
+>>>>>>> fb954c0 (Add `indepsym` wherever missing)
                      typeof(_colorvec), typeof(sys)}(f, g, mass_matrix,
                                                      analytic, tgrad, jac,
                                                      jvp, vjp, jac_prototype,
                                                      sparsity, Wfact,
                                                      Wfact_t,
                                                      paramjac, ggprime, syms,
+<<<<<<< HEAD
+=======
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:3329]: Merge error

**Comment by AayushSabharwal** [comment]
I honestly have no idea how that happened, they don't even show up as conflicts on VSCode. I'll get this right eventually

File: src/scimlfunctions.jl [hunk 10]
```diff
@@ -2118,19 +2214,22 @@ function DiscreteFunction(f; kwargs...)
...
-        DiscreteFunction{isinplace(f), recompile, Any, Any,
-                         Any, Any, Any}(newf, f.analytic, f.syms,
-                                        f.observed, f.sys)
+        DiscreteFunction{isinplace(f), recompile, typeof(ff), typeof(f.analytic),
+                         typeof(f.syms), typeof(f.indepsym), typeof(f.paramsyms),
+                         typeof(f.observed), typeof(f.sys)}(ff, f.analytic, f.syms,
+                                                            f.indepsym, f.paramsyms,
+                                                            f.observed, f.sys)
     else
-        DiscreteFunction{isinplace(f), recompile, typeof(newf), typeof(f.analytic),
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:2129]: Why was this reverted?

File: src/scimlfunctions.jl [hunk 11]
```diff
@@ -2189,30 +2291,58 @@ function SDEFunction{iip, recompile}(f, g;
...
+    Wfactiip = Wfact !== nothing ? isinplace(Wfact, 5, "Wfact", iip) : iip
+    Wfact_tiip = Wfact_t !== nothing ? isinplace(Wfact_t, 5, "Wfact_t", iip) : iip
+    paramjaciip = paramjac !== nothing ? isinplace(paramjac, 4, "paramjac", iip) : iip
+
+    nonconforming = (giip, jaciip, tgradiip, jvpiip, vjpiip, Wfactiip, Wfact_tiip,
+                     paramjaciip) .!= iip
+    if any(nonconforming)
+        nonconforming = findall(nonconforming)
+        functions = ["g", "jac", "tgrad", "jvp", "vjp", "Wfact", "Wfact_t", "paramjac"][nonconforming]
+        throw(NonconformingFunctionsError(functions))
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:2336]: What's going on here?

File: src/scimlfunctions.jl [hunk 12]
```diff
@@ -2747,7 +3117,112 @@ function DynamicalDDEFunction{iip}(f1, f2; kwargs...) where {iip}
 end
 DynamicalDDEFunction(f::DynamicalDDEFunction; kwargs...) = f
 
-function SDDEFunction{iip, recompile}(f, g;
+function SDDEFunction{iip, true}(f, g;
```
Comment (ChrisRackauckas) [inline src/scimlfunctions.jl:3120]: This is wrong... again.

**Comment by ChrisRackauckas** [comment]
This has so many rebase issues that it's just dangerous to even consider. Just start from a new branch on master.
